packages/components/src/navigator/navigator-provider/component.tsx

```diff
@@ -67,6 +139,9 @@ function UnconnectedNavigatorProvider(
 					...prevLocationHistory[ prevLocationHistory.length - 2 ],
 					isBack: true,
 					hasRestoredFocus: false,
+					restoreFocusTo:
```

Comment:
Instead of assigning `focusTargetSelector` in the `goTo` function  to the new location and then passing it to the previous location in `goBack` (as a new `restoreFocusTo` prop), could we instead assign it to the previous location directly in the `goTo` function?
This way we could avoid adding the `restoreFocusTo` prop altogether.

<details>
<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/navigator/navigator-provider/component.tsx b/packages/components/src/navigator/navigator-provider/component.tsx
index f025745690..84ffe9a382 100644
--- a/packages/components/src/navigator/navigator-provider/component.tsx
+++ b/packages/components/src/navigator/navigator-provider/component.tsx
@@ -36,7 +36,7 @@ import type {
 } from '../types';
 import { patternMatch } from '../utils/router';
 
-type MatchedPath = { params: object; id: string } | false;
+type MatchedPath = ReturnType< typeof patternMatch >;
 type ScreenAction = { type: string; screen: NavigatorScreenType };
 
 function screensReducer(
@@ -115,15 +115,33 @@ function UnconnectedNavigatorProvider(
 
 	const goTo: NavigatorContextType[ 'goTo' ] = useCallback(
 		( path, options = {} ) => {
-			setLocationHistory( ( prevLocationHistory ) => [
-				...prevLocationHistory,
-				{
-					...options,
+			setLocationHistory( ( prevLocationHistory ) => {
+				const { focusTargetSelector, ...restOptions } = options;
+
+				const newLocation = {
+					...restOptions,
 					path,
 					isBack: false,
 					hasRestoredFocus: false,
-				},
-			] );
+				};
+
+				if ( prevLocationHistory.length < 1 ) {
+					return [ newLocation ];
+				}
+
+				return [
+					...prevLocationHistory.slice( 0, -1 ),
+					// Assign `focusTargetSelector` to the previous location in history
+					// (the one we just navigated from).
+					{
+						...prevLocationHistory[
+							prevLocationHistory.length - 1
+						],
+						focusTargetSelector,
+					},
+					newLocation,
+				];
+			} );
 		},
 		[]
 	);
@@ -139,9 +157,6 @@ function UnconnectedNavigatorProvider(
 					...prevLocationHistory[ prevLocationHistory.length - 2 ],
 					isBack: true,
 					hasRestoredFocus: false,
-					restoreFocusTo:
-						prevLocationHistory[ prevLocationHistory.length - 1 ]
-							.focusTargetSelector,
 				},
 			];
 		} );
diff --git a/packages/components/src/navigator/navigator-screen/component.tsx b/packages/components/src/navigator/navigator-screen/component.tsx
index ef419f2a3c..2180681e72 100644
--- a/packages/components/src/navigator/navigator-screen/component.tsx
+++ b/packages/components/src/navigator/navigator-screen/component.tsx
@@ -116,9 +116,9 @@ function UnconnectedNavigatorScreen(
 
 		// When navigating back, if a selector is provided, use it to look for the
 		// target element (assumed to be a node inside the current NavigatorScreen)
-		if ( location.isBack && location?.restoreFocusTo ) {
+		if ( location.isBack && location?.focusTargetSelector ) {
 			elementToFocus = wrapperRef.current.querySelector(
-				location.restoreFocusTo
+				location.focusTargetSelector
 			);
 		}
 
@@ -137,7 +137,7 @@ function UnconnectedNavigatorScreen(
 		isInitialLocation,
 		isMatch,
 		location.isBack,
-		location.restoreFocusTo,
+		location.focusTargetSelector,
 	] );
 
 	const mergedWrapperRef = useMergeRefs( [ forwardedRef, wrapperRef ] );
diff --git a/packages/components/src/navigator/types.ts b/packages/components/src/navigator/types.ts
index 4dbeb99143..41d09d91ac 100644
--- a/packages/components/src/navigator/types.ts
+++ b/packages/components/src/navigator/types.ts
@@ -12,7 +12,6 @@ export type NavigatorLocation = NavigateOptions & {
 	isBack?: boolean;
 	path?: string;
 	hasRestoredFocus?: boolean;
-	restoreFocusTo?: string;
 };
 
 export type NavigatorContext = {
```

</details>

Reply:
yeah I actually thought about this, and was lazy enough that I didn't do it. But yeah I see the value in removing the duplicate keys.